sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs

```diff
@@ -0,0 +1,239 @@
+            }
+
+            public bool SupportsPartialResults => true;
+
+            public IEnumerable<Diagnostic> GetDiagnostics() =>
+                nodes.Distinct().Select(node => Diagnostic.Create(rule, node.GetLocation()));
+
+            public void Dispose()
+            {
+                // Nothing to do here.
+            }
+
+            private void AddNode(SyntaxNode syntaxNode) => nodes.Add(syntaxNode);
+        }
+
+        private sealed class SerializationBinderCheck : ExplodedGraphCheck
+        {
+            private const int BindToTypeParameterCount = 2;
+
+            private static readonly ImmutableArray<KnownType> typesWithBinder =
+                ImmutableArray.Create(
+                    KnownType.System_Runtime_Serialization_Formatters_Binary_BinaryFormatter,
+                    KnownType.System_Runtime_Serialization_NetDataContractSerializer,
+                    KnownType.System_Runtime_Serialization_Formatters_Soap_SoapFormatter,
+                    KnownType.System_Web_UI_ObjectStateFormatter);
+
+            private readonly Dictionary<ITypeSymbol, bool> binderValidityMap = new Dictionary<ITypeSymbol, bool>();
+
+            private readonly Action<SyntaxNode> addNode;
+
+            public SerializationBinderCheck(AbstractExplodedGraph explodedGraph, Action<SyntaxNode> addNode)
+                : base(explodedGraph)
+            {
+                this.addNode = addNode;
+            }
+
+            public override ProgramState ObjectCreated(ProgramState programState, SymbolicValue symbolicValue,
+                SyntaxNode instruction)
+            {
+                if (instruction is ObjectCreationExpressionSyntax objectCreation)
+                {
+                    var typeSymbol = semanticModel.GetTypeInfo(instruction).Type;
+                    if (IsFormatterWithBinder(typeSymbol) &&
+                        !HasSafeBinderInitialization(objectCreation.Initializer))
+                    {
+                        programState = programState.SetConstraint(symbolicValue, SerializationBinderConstraint.Unsafe);
+                    }
+                }
+
+                return base.ObjectCreated(programState, symbolicValue, instruction);
+            }
+
+            public override ProgramState PreProcessInstruction(ProgramPoint programPoint, ProgramState programState)
+            {
+                var instruction = programPoint.Block.Instructions[programPoint.Offset];
+
+                programState = instruction switch
+                {
+                    MemberAccessExpressionSyntax memberAccess => VisitMemberAccess(memberAccess, programState),
+                    AssignmentExpressionSyntax assignmentExpressionSyntax => VisitAssignmentExpression(assignmentExpressionSyntax, programState),
+                    _ => programState
+                };
+
+                return base.PreProcessInstruction(programPoint, programState);
```

Comment:
`return programState` would have the exact same behaviour and would be clearer in what it does, as otherwise we could think there is some logic done by the parent class, which is not the case